### ISSUE TITLE
eigen package

### ISSUE THREAD

**Issue by mhscott** [OpenSees/OpenSees#1401]
@jaabell @MassimoPetracca Can you point me to where you actually use the eigen package in the `SRC/material/nD` directory? Seems a bit much for constitutive models and majorly affects the current build (moving to all cmake is slow going and we won't ween off `make` or Visual Studio any time soon). And are the licensing terms of the eigen library compatible with OpenSees?

**Comment by MassimoPetracca** [comment]
Hi @mhscott .

> Can you point me to where you actually use the eigen package in the SRC/material/nD directory?

Only in the subfolder SRC/material/nD/ASDPlasticMaterial

> And are the licensing terms of the eigen library compatible with OpenSees?

Yes. Eigen is MPL v.2. (see the FAQ [here](https://www.mozilla.org/en-US/MPL/2.0/FAQ/))
It's a weak copyleft and can be used also in derivative works under proprietary licenses, or under strong copyleft licenses as GPL (see Q8, Q14).
It also does not pose any issue with web-based services according to their definition of "distribute" (Q17).
The only requirement is to make available any changes made to the MPL2 licensed files (which is not the case here since we do not even include them in the repo, it's either conan that installs it or the git-submodule for old VS or makes)

> Seems a bit much for constitutive models and majorly affects the current build (moving to all cmake is slow going and we won't ween off make or Visual Studio any time soon).

Sure, you're right. We forgot to update the VS project files.
I'll do a PR soon. It's just a matter of adding 2 include-dirs and a cpp file in the material project file.

The last point is to explain to the user that now instead of only this:
```
git clone https://github.com/OpenSees/OpenSees.git
```
they need to
```
git clone https://github.com/OpenSees/OpenSees.git
git submodule init
git submodule update 
```

It should be stated somewhere in the session 2.1.2. Obtaining OpenSees Source Code of the build doc.

@mhscott  what do you suggest? Do I proceed with the PR for the material vs project file and assume the user uses the git submodule features for the old-style VS and make files?

**Comment by mhscott** [comment]
Thanks. This all seems like a lot of changes _for a constitutive model_. May be worth looking into an `#ifdef` or keeping a separate public branch in the OpenSees GitHub repo.

**Comment by MassimoPetracca** [comment]
@mhscott ok so we can do something like this.
We can use a preprocessor macro to activate the compilation of ASD plastic material. Then we will automatically define this macro in cmake if Conan manages to provide eigen.
For all other build tools (plain vs or make) it will not be defined. 
Does it sound good?

**Comment by mhscott** [comment]
Sounds good. Thanks!

**Comment by jaabell** [comment]
Hola querido @mhscott 

I'd like to mention you that this issue was previously discussed with @fmckenna via email on August 31, 2022. We anticipated the impact of introducing this new dependency on the OpenSees build process, and @fmckenna approved it based on the merits of the contribution.

`ASDPlasticMaterial ` is not _merely a single constitutive model_; it represents an **entire family of plasticity models**. Currently, it includes approximately 60 models, and this number is expected to grow combinatorially with the introduction of new components such as yield surfaces and hardening functions.

Eigen is utilized as a foundational tensor class for these material models. It's a lightweight, header-only library that doesn't require compilation of new dependencies. Eigen and  `ASDPlasticMaterial` employ extensive template metaprogramming, providing both modularity and performance, while also facilitating expressive programming techniques. It is _perfectly suited_ for the task of creating this particular family of constitutive models.  

Furthermore, integrating Eigen as a dependency in OpenSees allows its use in elements, enabling the creation of new families of elements analogous to the `ASDPlasticMaterial` family of constitutive models, among other applications. We have plans for similar contributions in the near future. 

Saludos desde Chile!

**Comment by mhscott** [comment]
Thanks for the explanation. Late 2022 there also was a plan to remove Makefiles and Visual Studio files, but that hasn't happened. Things move slowly. Until the issues with compiling OpenSees with cmake and conan are resolved, we should keep the status quo and Massimo's workaround achieves that outcome.

**Comment by MassimoPetracca** [comment]
fixed in PR OpenSees/OpenSees#1424

**Comment by mhscott** [comment]
Thanks, @MassimoPetracca!
